### ISSUE TITLE
Add DllSafeSearchPathAttribute by default to all .NET Core assemblies

### ISSUE THREAD

**Issue by vatsan-madhavan** [dotnet/arcade#1888]
Can we consider adding the following attribute to all .NET Core assemblies by default. 

```
[assembly:DefaultDllImportSearchPathsAttribute(DllImportSearchPath::AssemblyDirectory | DllImportSearchPath::System32)]
```

BuildTools already seems to do this here: 

- Initializes `IncludeDllSafeSearchPathAttribute` property [here](https://github.com/dotnet/buildtools/blob/4642e67685b9c21c4d5160ea7665843a48ed8155/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets#L43)
- Generates the assembly attribute [here](https://github.com/dotnet/buildtools/blob/6736870b84e06b75e7df32bb84d442db1b2afa10/src/Microsoft.DotNet.Build.Tasks/PackageFiles/GenerateAssemblyInfo.Sdk.targets#L41)

Assemblies that are now generated in an Arcade SDK based repo seem to be silently losing this assembly level attribute. 

cc @ericstj, @rladuca, @stevenbrix, @dotnet/wpf-developers

**Comment by vatsan-madhavan** [comment]
I don't think all .NET Core projects need to do this, but this is the right policy for CoreFX libraries. 

My request (at least the very first line) is not well phrased, but it is in fact only a request to add this attribute _conditionally_ - when `IncludeDllSafeSearchPathAttribute` property is set to `true`. It would make sense for CoreFX repos to set this property to `true`.

**Comment by tmat** [comment]
The .NET Core SDK allows to add assembly attributes via `AssemblyAttribute` item group.
So the repos that want this policy could add this to their `Directory.Build.props` file:

```xml
<ItemGroup>
  <AssemblyAttribute Include="DefaultDllImportSearchPathsAttribute" _Parameter1="DllImportSearchPath::AssemblyDirectory | DllImportSearchPath::System32" />
<ItemGroup>
```

Given that I don't see much benefit of adding `IncludeDllSafeSearchPathAttribute` property.

**Comment by vatsan-madhavan** [comment]
Got it - I wasn't aware of this - thanks!

**Comment by vatsan-madhavan** [comment]
Turns out that this is not really available through the .NET Core Sdk. Attributes that require literals as parameters aren't quite usable through `AssemblyAttribute` - see https://github.com/Microsoft/msbuild/issues/2281. 

Can we reconsider provisioning this through the Arcade SDK ?

**Comment by tmat** [comment]
You can also include a .cs file to all projects that contains the attribute (or any number of attributes that should be set in all projects)

**Comment by ericstj** [comment]
See https://github.com/dotnet/corefx/blob/49d249fa023e8575dba7e35febb4446e4fa13f98/eng/versioning.targets#L66 for how we do it in CoreFx.

**Comment by tmat** [comment]
I don't think extra target is necessary since it still needs to special case each attribute. Including .cs file that declares all attributes needed is much simpler.

**Comment by ericstj** [comment]
Agreed that a single source file is simpler. Just wanted to share the corefx solution which was handling a more generic problem.

**Comment by markwilkie** [comment]
@vatsan-madhavan - next steps?

**Comment by vatsan-madhavan** [comment]
If projects want to customize their policies, and a repo used multiple languages (C#, VB, C++/CLI), it is better to depend on generated code over a single (i.e.,, a smattering of) source(s).

== 

For now, I've built a custom target to solve this - I think this issue can be closed. 

== 

Unless...

The target I've built is a reasonable replacement for `AssemblyAttribute` that can be used in place of, or as a supplement to it. I'll be happy to PR it into Arcade if there is interest, so others could benefit as well. 

The usage looks about like this and can handle an arbitrary number of literal or non-literal parameters. 

```
 Input: 
    
      An ItemGroup containing assembly-level Attribute information. 
      Typical format for ExtendedAssemblyInfo looks like this: 
    
      <ItemGroup>
        <ExtendedAssemblyInfo Include="AttributeType">
          <Parameter1>true</Parameter1>
          <IsLiteral1>true</IsLiteral1>
          
          <Parameter2>https://www.microsoft.com</Parameter2>
        </ExtendedAssemblyInfo>
      </ItemGroup>
      
        Attribute Type:    The type (full type name) is specified as the identity element (Include=)
        Parameters:        Parameters are specified in metadata values Parameter1, Parameter2, Parameter3, etc. 
                            Parameters are treated as strings by default. 
                            Literal parameters (for e.g., numbers, boolean literals, enums etc.) should be indicated as IsLiteral=true corresponding to the parameter position. 
                            For e.g., for Parameter6, indicate it as <IsLiteral6>true</IsLiteral>                     
```

**Comment by markwilkie** [comment]
Seems like this would be a nice PR.  If you wouldn't mind that'd be great @vatsan-madhavan.  Thanks!
